test/e2e/certificate_publisher_test.go

```diff
@@ -0,0 +1,168 @@
+// +build e2e
+
+package e2e
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	ingresscontroller "github.com/openshift/cluster-ingress-operator/pkg/operator/controller"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+// TestCreateIngressControllerThenSecret creates an ingresscontroller that
+// references a secret that does not exist, then creates the secret and verifies
+// that the operator updates the "router-certs" global secret.
+func TestCreateIngressControllerThenSecret(t *testing.T) {
+	cl, ns, err := getClient()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Create the ingresscontroller.
+	var one int32 = 1
+	ic := &operatorv1.IngressController{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-ingresscontroller-then-secret",
+			Namespace: ns,
+		},
+		Spec: operatorv1.IngressControllerSpec{
+			DefaultCertificate: &corev1.LocalObjectReference{
+				Name: "e2e-test-secret",
+			},
+			Domain: "test-ingresscontroller-then-secret",
+			EndpointPublishingStrategy: &operatorv1.EndpointPublishingStrategy{
+				Type: operatorv1.PrivateStrategyType,
+			},
+			Replicas: &one,
+		},
+	}
+	if err := cl.Create(context.TODO(), ic); err != nil {
+		t.Fatalf("failed to create the ingresscontroller: %v", err)
+	}
+	defer func() {
+		if err := cl.Delete(context.TODO(), ic); err != nil {
+			t.Fatalf("failed to delete the ingresscontroller: %v", err)
+		}
+	}()
+
+	// Wait for the deployment to exist.
+	err = wait.PollImmediate(1*time.Second, 60*time.Second, func() (bool, error) {
+		deployment := &appsv1.Deployment{}
+		if err := cl.Get(context.TODO(), types.NamespacedName{Namespace: "openshift-ingress", Name: "router-test-ingresscontroller-then-secret"}, deployment); err != nil {
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		t.Fatalf("failed to get deployment: %v", err)
+	}
+
+	// Create the secret.
+	err, secret := createDefaultCertTestSecret(cl)
+	if err != nil {
+		t.Fatalf("failed to create secret: %v", err)
+	}
+	defer func() {
+		if err := cl.Delete(context.TODO(), secret); err != nil {
+			t.Errorf("failed to delete secret: %v", err)
+		}
+	}()
+
+	// Wait for the "router-certs" secret to be updated.
+	err = wait.PollImmediate(1*time.Second, 60*time.Second, func() (bool, error) {
+		globalSecret := &corev1.Secret{}
+		if err := cl.Get(context.TODO(), types.NamespacedName{Namespace: ingresscontroller.GlobalMachineSpecifiedConfigNamespace, Name: "router-certs"}, globalSecret); err != nil {
```

Comment:
Should get the name from a function in `names.go`?

Reply:
I've been avoiding using internal functions in the tests, but I can do that.

Reply:
if the e2e test knows specifics how we're laying out and naming resources, then the name information isn't really internal to the test, at least

Reply:
Fixed.  Thanks!